Add functions to split a robot model into submodel objects

isorrentino opened the pull request:
cc @GiulioRomualdi @S-Dafarra

S-Dafarra commented on src/Estimators/include/BipedalLocomotion/RobotDynamicsEstimator/SubModelCreator.h:
```diff
@@ -0,0 +1,176 @@
+ * distributed under the terms of the BSD-3-Clause license.
+ */
+
+#ifndef SUBMODELCREATOR_H
+#define SUBMODELCREATOR_H
```
Can you prepend ``BIPEDAL_LOCOMOTION_`` to the guard?

S-Dafarra commented on src/Estimators/include/BipedalLocomotion/RobotDynamicsEstimator/SubModelCreator.h:
```diff
@@ -0,0 +1,176 @@
+ */
+struct FT : Sensor
+{
+    enum class Direction : int { positive=1, negative=-1, notspecified=0 };
```
I realized that we don't have a specific code style for the components of an ``enum class``, although the most common style we used is with the first letter capitol and then camel case. @GiulioRomualdi what do you think?

S-Dafarra commented on src/Estimators/include/BipedalLocomotion/RobotDynamicsEstimator/SubModelCreator.h:
```diff
@@ -0,0 +1,176 @@
+ * @param subModel iDynTree Model describing one of the sub-models.
+ * @return a vector of integers describing the index of the joint in the full model.
+ */
+std::vector<int> createJointMapping(const iDynTree::Model fullModel, const iDynTree::Model subModel);
```
I would suggest passing the models by reference to avoid copies.

S-Dafarra commented on src/Estimators/include/BipedalLocomotion/RobotDynamicsEstimator/SubModelCreator.h:
```diff
@@ -0,0 +1,176 @@
+ * @brief createJointMapping creates a map between the joint indeces in the submodel and the same joints in the full model.
+ * @param fullModel iDynTree Model corresponging to the full robot model.
+ * @param subModel iDynTree Model describing one of the sub-models.
+ * @return a vector of integers describing the index of the joint in the full model.
```
Which joint?

S-Dafarra commented on src/Estimators/tests/RobotDynamicsEstimator/ConfigFolderPath.h.in:
```diff
@@ -0,0 +1,18 @@
+    return std::string(SOURCE_CONFIG_DIR) + "/model.ini";
+}
+
+#endif // CONFIG_FOLDERPATH_H_IN
```
We have a similar file here:https://github.com/ami-iit/bipedal-locomotion-framework/blob/dd649fe39767141e32baa85060b9198fbf714c43/src/ParametersHandler/tests/ConfigFolderPath.h.in

I wonder if we can manage to avoid duplicates. What do you think @GiulioRomualdi?

GiulioRomualdi replied:
We may think to have it in the cmake folder. 
I think we can consider it in a separate PR, do you agree?

S-Dafarra replied:
No problem for me

isorrentino commented:
Hi @S-Dafarra I addressed all the comments besides https://github.com/ami-iit/bipedal-locomotion-framework/pull/604#discussion_r1104144580 and https://github.com/ami-iit/bipedal-locomotion-framework/pull/604#discussion_r1104158771. I will wait for @GiulioRomualdi's answers. Thanks for the review.

S-Dafarra commented:
As a more general comment, I have seen that you removed the ``RobotDynamicsEstimator`` namespace. Now all the ``SubModel`` related functions are in the "general" namespace ``Estimators``.
It is not a big deal, but I would expect the symbols in the general ``Estimators`` namespace to be related to an estimator, while in this case are functions related to the use of a sub-model.
One possibility would be to move those functions within the struct ``SubModel``.
Another, trickier, possibility is to have a more class-oriented approach. I have noticed that many of those methods require a submodel as input (all the ``attach`` functions). This can be avoided if the function is a class method. The functions that instead create a ``SubModel`` can be a ``static`` method. In other words, we could have
```cpp
class SubModel {
    //These below are all private. If we need to expose them, we could use appropriate getters/setters
    iDynTree::Model model; /**< iDynTree Model object describing the submodel */
    std::vector<int> jointListMapping; /**< Each element contains an index describing the joint position in the full model*/
    std::vector<FT> ftList; /**< List of force/torque sensors in the submodel */
    std::vector<Sensor> accelerometerList; /**< List of accelerometers in the submodel */
    std::vector<Sensor> gyroscopeList; /**< List of gyroscopes in the submodel */
    std::vector<std::string> externalContactList; /**< List of the additional external contacts */
public:
    static bool getSubModels(iDynTree::ModelLoader & mdlLdr,
                  iDynTree::KinDynComputations & kinDyn,
                  const std::vector<BipedalLocomotion::Estimators::Sensor> & ftSensorList,
                  const std::vector<BipedalLocomotion::Estimators::Sensor> & accList,
                  const std::vector<BipedalLocomotion::Estimators::Sensor> & gyroList,
                  const std::vector<std::string> & externalContacts,
                  std::vector<BipedalLocomotion::Estimators::SubModel> & subModelList);
    /* ...... */
    std::vector<std::string> attachExternalContactsToSubModel(const std::vector<std::string> & contactsFromConfig); //note that I am not passing a submodel here
};
```

Anyhow, these are just possible suggestions. Feel free to ignore them :wink:

isorrentino commented:
@S-Dafarra @GiulioRomualdi I completely changed the file structure. I created two classes, one for splitting the model into sub-models and one that represents the sub-model object (what before was a struct). All members are private and are only accessible via get methods. I have also added back the `RobotDynamicsEstimator` namespace.

S-Dafarra commented on src/Estimators/include/BipedalLocomotion/RobotDynamicsEstimator/SubModel.h:
```diff
@@ -0,0 +1,256 @@
+    * @brief Access model.
+    * @return The model of the SubModel.
+    */
+    iDynTree::Model getModel();
+
+    /**
+     * @brief Access jointListMapping.
+     * @return the mapping between the joint indeces in the sub-model and the joint indeces in the full-model.
+     */
+    std::vector<int> getJointMapping();
+
+    /**
+     * @brief Access ftList.
+     * @return the list of FT objects which is the list of force/torque sensors.
+     */
+    std::vector<FT> getFTList();
+
+    /**
+     * @brief Access accelerometerList.
+     * @return a list of Sensor objects describing the accelerometers contained in the sub-model.
+     */
+    std::vector<Sensor> getAccelerometerList();
+
+    /**
+     * @brief Access gyroscopeList.
+     * @return a list of Sensor objects describing the gyroscope contained in the sub-model.
+     */
+    std::vector<Sensor> getGyroscopeList();
+
+    /**
+     * @brief Access externalContactList.
+     * @return a list of strings describing frame names of the external contacts for the sub-model.
+     */
+    std::vector<std::string> getExternalContactList();
```
Here you can output all these by const reference and make the methods const, e.g.
```cpp
const iDynTree::Model& getModel() const;
```

The output const reference avoids creating a temporary object every time the method is called. In fact, right now, it creates a brand new model/vector at every call. This can be problematic if the output requires a lot of memory.

The ``const`` after the method name is to notify that the method can be called even when you have a const reference to the class, because you are making explicit the fact that by calling the method, the internal state does not change.

S-Dafarra commented on src/Estimators/include/BipedalLocomotion/RobotDynamicsEstimator/SubModel.h:
```diff
@@ -0,0 +1,256 @@
+ */
+struct FT : Sensor
+{
+    enum class Direction : int { positive=1, negative=-1, notspecified=0 };
```
Just adding again the comment on the style of the ``enum`` class.

S-Dafarra commented on src/Estimators/src/SubModel.cpp:
```diff
@@ -0,0 +1,409 @@
+
+void blfEstim::SubModelCreator::setModel(const iDynTree::Model & model)
+{
+    this->model = model.copy();
```
Here in principle you can simply use 
```
this->model = model;
```
because in C++ the copy operator is called https://github.com/robotology/idyntree/blob/988c915f82a485fb44925a5a1e80543db75fcf7a/src/model/src/Model.cpp#L81-L90

Right now you are making an extra copy since the ``model.copy()`` generates a temporary model that is then copied in ``this->model`` via the same copy operator I mentioned above, and then destroyed immediately afterward.

Also, to avoid confusion to a reader, it is often preferable to have some naming convention for the private members, like ``m_model``. In this way, 
```cpp
m_model = model;
```
clearly copies an input object to a private class member.

S-Dafarra reviewed (commented):
Good job in restructuring the code. I believe it is much tidier now.

Here some suggestions to avoid performance drops

isorrentino commented:
@S-Dafarra thanks again. I modified the code following your suggestions.

S-Dafarra reviewed (approved):
Just as a final comment, you can still mark the getters of ``SubModel`` as ``const`` (e.g. ``const iDynTree::Model& getModel() const;``. This might be helpful if you want to access those methods when having a const reference to ``SubModel``.

Anyhow, none of the remaining comments is blocking for me.

isorrentino commented:
> Anyhow, none of the remaining comments is blocking for me.

Great! Thanks again, you gave me really useful tips :)

GiulioRomualdi commented on src/Estimators/include/BipedalLocomotion/RobotDynamicsEstimator/SubModel.h:
```diff
@@ -0,0 +1,256 @@
+ * specified as configuration parameters.
+ */
+
+class SubModelCreator {
+
+private:
```
```suggestion
class SubModelCreator
{
```

GiulioRomualdi commented on src/Estimators/include/BipedalLocomotion/RobotDynamicsEstimator/SubModel.h:
```diff
@@ -0,0 +1,256 @@
+    std::shared_ptr<iDynTree::KinDynComputations> m_kinDyn; /**< Pointer to a KinDynComputations object */
+    std::vector<SubModel> m_subModelList; /**< List of SubModel objects */
+
+private:
```
```suggestion
```

GiulioRomualdi commented on src/Estimators/src/SubModel.cpp:
```diff
@@ -0,0 +1,409 @@
+namespace blfEstim = BipedalLocomotion::Estimators::RobotDynamicsEstimator;
+
+
+bool blfEstim::SubModelCreator::splitModel(const std::vector<std::string> & ftFrameList, std::vector<iDynTree::Model> & idynSubModels)
```
could you please apply the clang-format?

GiulioRomualdi commented on src/Estimators/src/SubModel.cpp:
```diff
@@ -0,0 +1,409 @@
+        }
+        else
+        {
+            std::string firstLink = sensorFTFromModel->getFirstLinkName();
+            std::string secondLink = sensorFTFromModel->getSecondLinkName();
```
```suggestion
            const std::string firstLink = sensorFTFromModel->getFirstLinkName();
            const std::string secondLink = sensorFTFromModel->getSecondLinkName();
```

GiulioRomualdi reviewed (commented):
Could you also update the CHANGELOG?

traversaro commented on src/Estimators/include/BipedalLocomotion/RobotDynamicsEstimator/SubModel.h:
```diff
@@ -0,0 +1,273 @@
+{
+    enum class Direction : int
+    {
+        Positive = 1,
```
Probably it may be worth to specify if "positive" means the force that the rest of the system applied on the considered submodel, or that the submodel applies on the rest of the system. "Positive" or "Negative" on their own risk of being a bit ambigous.

isorrentino replied:
> Probably it may be worth to specify if "positive" means the force that the rest of the system applied on the considered submodel, or that the submodel applies on the rest of the system. "Positive" or "Negative" on their own risk of being a bit ambiguous.

@traversaro  Yes, it's ambiguous. Are `Compressing` and `Stretching` more explicative? Of course, I'm going to add also the documentation as you suggested.

traversaro replied:
I don't think `Compressing` or `Stretching` are more explicative, as depending on the sign of the measure you can have a sensor that is marked as "Compressing" that is measuring a force that is "Stretching" and the other way around, but this is just my taste.

isorrentino commented:
@GiulioRomualdi the PR is again ready for review.

GiulioRomualdi commented:
Thank you @isorrentino I will squash and merge the PR :)
